pkg/workflow/workflow.go

```diff
@@ -500,64 +516,134 @@ func (e *engine) todoByIndex(taskRunners []wfTypes.TaskRunner) []wfTypes.TaskRun
+			debugContext := debug.NewContext(e.cli, e.rk, e.app, step)
+			if err := debugContext.Set(v); err != nil {
+				return err
 			}
-			continue
+			return nil
 		}
-
-		if val, exists := wfCtx.GetValueInMemory(wfTypes.ContextPrefixBackoffReason, status.ID); !exists || val != status.Message {
-			wfCtx.SetValueInMemory(status.Message, wfTypes.ContextPrefixBackoffReason, status.ID)
-			wfCtx.DeleteValueInMemory(wfTypes.ContextPrefixBackoffTimes, status.ID)
+	}
+	status, operation, err := runner.Run(wfCtx, options)
+	if err != nil {
+		return true, err
+	}
+	if parentStepName != "" {
+		subStepStatus := common.WorkflowSubStepStatus{
+			ID:               status.ID,
+			Name:             status.Name,
+			Type:             status.Type,
+			Phase:            status.Phase,
+			Message:          status.Message,
+			Reason:           status.Reason,
+			FirstExecuteTime: status.FirstExecuteTime,
+			LastExecuteTime:  status.LastExecuteTime,
 		}
-		wfCtx.IncreaseCountValueInMemory(wfTypes.ContextPrefixBackoffTimes, status.ID)
+		e.updateSubStepStatus(parentStepName, subStepStatus)
+	} else {
+		e.updateStepStatus(status)
+	}
+
+	e.failedAfterRetries = e.failedAfterRetries || operation.FailedAfterRetries
+	e.waiting = e.waiting || operation.Waiting
+	if status.Phase == common.WorkflowStepPhaseSucceeded || (status.Phase == common.WorkflowStepPhaseRunning && status.Type == wfTypes.WorkflowStepTypeSuspend) {
+		wfCtx.DeleteValueInMemory(wfTypes.ContextPrefixBackoffTimes, status.ID)
+		wfCtx.DeleteValueInMemory(wfTypes.ContextPrefixBackoffReason, status.ID)
 		if err := wfCtx.Commit(); err != nil {
-			return errors.WithMessage(err, "commit workflow context")
+			return true, errors.WithMessage(err, "commit workflow context")
 		}
-		if e.isDag() {
-			continue
+
+		e.finishStep(operation)
+		if e.needStop() {
+			return true, nil
 		}
-		e.checkFailedAfterRetries()
-		return nil
+		return false, nil
 	}
-	return nil
+
+	if val, exists := wfCtx.GetValueInMemory(wfTypes.ContextPrefixBackoffReason, status.ID); !exists || val != status.Message {
+		wfCtx.SetValueInMemory(status.Message, wfTypes.ContextPrefixBackoffReason, status.ID)
+		wfCtx.DeleteValueInMemory(wfTypes.ContextPrefixBackoffTimes, status.ID)
+	}
+	wfCtx.IncreaseCountValueInMemory(wfTypes.ContextPrefixBackoffTimes, status.ID)
+	if err := wfCtx.Commit(); err != nil {
+		return true, errors.WithMessage(err, "commit workflow context")
+	}
+	if dag {
+		return false, nil
+	}
+	e.checkFailedAfterRetries()
+	return true, nil
+}
+
+func (e *engine) stepWithSubSteps(runner wfTypes.TaskRunner) (bool, error) {
+	stepName := runner.Name()
+	status, _, err := runner.Run(e.wfCtx, &wfTypes.TaskRunOptions{})
+	if err != nil {
+		return true, err
+	}
+	status.SubStepsStatus = e.getStepStatus(stepName).SubStepsStatus
+	e.updateStepStatus(status)
+
+	todoSubTaskRunners, pendingSubTaskRunners := e.getTodoSubTaskRunners(runner)
+	for _, todoSubTaskRunner := range todoSubTaskRunners {
+		// subStep only use dag mode
```

Comment:
Maybe it's better to add `mode` in subSteps and make it configurable. /cc @yangsoon